File: solutions/elixir/bob/2/lib/bob.ex
defmodule Bob do
  def hey(input) do
    input = String.trim(input)

    # 1. match all non lowercase
    # 2. match first uppercase
    # 3. match all non lowercase
    yelling? = String.match?(input, ~r/^[^\p{Ll}]*\p{Lu}[^\p{Ll}]*$/u)

    question? =
      input
      |> String.trim_trailing()
      |> String.last() == "?"

    cond do
      input == "" ->
        "Fine. Be that way!"

      yelling? and question? ->
        "Calm down, I know what I'm doing!"

      yelling? ->
        "Whoa, chill out!"

      question? ->
        "Sure."

      true ->
        "Whatever."
    end
  end
end

File: solutions/haskell/bob/2/src/Bob.hs
module Bob (responseFor) where

import Data.Char (toUpper, toLower, isSpace)

import Data.List  (dropWhile, dropWhileEnd)

responseFor :: String -> String
responseFor xs
  | null text             = "Fine. Be that way!"
  | yellingP && questionP = "Calm down, I know what I'm doing!"
  | yellingP              = "Whoa, chill out!"
  | questionP             = "Sure."
  | otherwise             = "Whatever."
 where
  text      = (dropWhile isSpace . dropWhileEnd isSpace) xs
  uppercase = map toUpper text
  yellingP  = map toLower text /= uppercase && uppercase == text
  questionP = last text == '?'
